### PR TITLE
Allow setting custom finder for authenticators.

### LINKED ISSUE #7162: [Enhancement] Add a 'select' key to the AuthComponent config
I'm using the **AuthComponent** and i want to specify the **Users Table fields**. 
By default the **BaseAuthenticate::_query($username)** select all the fields from the Users Table:
```php
$query = $table->find('all')
            ->where($conditions);
```
We can't specify the fields to select in the AuthComponent config like this:
```php
$this->loadComponent('Auth', [
	// ...
    'authenticate' => [
        'Form' => [
             // eg. add the 'select' key to the config
            'select' => ['id', 'email', 'fullname', 'phone', 'fax', 'lastlogin', 'ip', 'role_id'],
            'fields' => ['username' => 'email', 'password' => 'password'],
            'scope' => ['Users.state' => 1, 'Users.deleted' => 0],
            'contain' => ['Roles' => ['fields' => ['id', 'name']]]
        ],
    ]
]);
```
So i override the **BaseAuthenticate::_query($username)** by creating a custom **MyFormAuthenticate::_query($username)** in the **src/Auth** folder like this:
```php
<?php

namespace App\Auth;

use Cake\Auth\FormAuthenticate;
use Cake\ORM\TableRegistry;

class MyFormAuthenticate extends FormAuthenticate
{

    /**
     * Get query object for fetching user from database.
     *
     * @param string $username The username/identifier.
     * @return \Cake\ORM\Query
     */
    protected function _query($username)
    {
        $config = $this->_config;
        $table = TableRegistry::get($this->_config['userModel']);

        $conditions = [$table->aliasField($config['fields']['username']) => $username];
        if ($config['scope']) {
            $conditions = array_merge($conditions, $config['scope']);
        }

        $query = $table->find('all')
            ->where($conditions);
        
        // I add this lines
        if($config['select']){
            $query = $query->select($config['select']);
        }

        if ($config['contain']) {
            $query = $query->contain($config['contain']);
        }

        return $query;
    }
}
```
I modified also the config of the AuthComponent like this:
```php
$this->loadComponent('Auth', [
	// ...
    'authenticate' => [
        // 'Form' to 'MyForm'
        'MyForm' => [
             // eg. add the 'select' key to the config
            'select' => ['id', 'email', 'fullname', 'phone', 'fax', 'lastlogin', 'ip', 'role_id'],
            'fields' => ['username' => 'email', 'password' => 'password'],
            'scope' => ['Users.state' => 1, 'Users.deleted' => 0],
            'contain' => ['Roles' => ['fields' => ['id', 'name']]]
        ],
    ]
]);
```
Now it works and i can specify the fields to select from the Users Table.

I think this config option must exist by default :)

### PR DESCRIPTION
**Pull request by ADmad**
Closes #7162

### CODE REVIEW

**Comment by bravo-kernel** [comment]
:bulb:

File: src/Auth/BaseAuthenticate.php [hunk 1]
```diff
@@ -37,6 +37,7 @@
      * - `userModel` The alias for users table, defaults to Users.
      * - `scope` Additional conditions to use when looking up and authenticating users,
      *    i.e. `['Users.is_active' => 1].`
+     * - `finder` The finder method to use to fetch user record. Defaults to 'all'.
```
Comment (markstory) [inline src/Auth/BaseAuthenticate.php]: We could probably deprecate the `contain` & `scope` options. The new `finder` option gives people more control/flexibility than those other options.

File: tests/TestCase/Auth/FormAuthenticateTest.php [hunk 2]
```diff
@@ -296,6 +301,32 @@ public function testPluginModel()
...
+            'userModel' => 'AuthUsers',
+            'finder' => 'auth'
+        ]);
+
+        $result = $this->auth->authenticate($request, $this->response);
+        $expected = [
+            'id' => 1,
+            'username' => 'mariano',
+        ];
+        $this->assertEquals($expected, $result, 'Result should not contain "created" and "modifield" fields');
```
Comment (markstory) [inline tests/TestCase/Auth/FormAuthenticateTest.php]: Typo in modified.

File: src/Auth/BaseAuthenticate.php (hunk 1 above)
Comment (ADmad) [inline src/Auth/BaseAuthenticate.php]: Yup I had thought that too. Though we should skip triggering errors like we normally do for deprecated features. Will just undocument those options and mention they are deprecated in migration guide.

Comment (markstory) [inline src/Auth/BaseAuthenticate.php]: > Though we should skip triggering errors like we normally do for deprecated features. Will just undocument those options and mention they are deprecated in migration guide.

Agreed.

Comment (Marlinc) [inline src/Auth/BaseAuthenticate.php]: I agree that deprecating the scope and contain parameters makes sense.

File: tests/TestCase/Auth/FormAuthenticateTest.php [hunk 3]
```diff
@@ -244,30 +250,12 @@ public function testAuthenticateIncludesVirtualFields()
...
-    }
-
-    /**
      * test a model in a plugin.
      *
      * @return void
      */
     public function testPluginModel()
     {
-        Cache::delete('object_map', '_cake_core_');
```
Comment (markstory) [inline tests/TestCase/Auth/FormAuthenticateTest.php]: Nice find.

**Comment by markstory** [comment]
Looks good to me.

**Comment by lorenzo** [comment]
@ADmad Can you document this with an example in the book?

**Comment by ADmad** [comment]
@lorenzo I had already documented the new option, made a new PR with example too.

**Pull request merged**